src/librustc/mir/interpret/value.rs

```diff
@@ -294,12 +294,35 @@ impl<'tcx, Tag> Scalar<Tag> {
         Scalar::Raw { data: f.to_bits() as u128, size: 8 }
     }
 
+    #[inline]
+    pub fn to_bits_or_ptr(
+        self,
+        target_size: Size,
+        cx: &impl HasDataLayout,
+    ) -> Result<u128, Pointer<Tag>> {
+        match self {
+            Scalar::Raw { data, size } => {
+                assert_eq!(target_size.bytes(), size as u64);
+                assert_ne!(size, 0, "to_bits cannot be used with zsts");
+                assert_eq!(truncate(data, target_size), data,
+                            "Scalar value {:#x} exceeds size of {} bytes", data, size);
+                Ok(data)
+            }
+            Scalar::Ptr(ptr) => {
+                assert_eq!(target_size, cx.data_layout().pointer_size);
+                Err(ptr)
+            }
+        }
+    }
+
     #[inline]
     pub fn to_bits(self, target_size: Size) -> EvalResult<'tcx, u128> {
         match self {
```

Comment:
deduplicate with `to_bits_or_ptr` by calling that function

Reply:
I'd like to, but that needs an `impl HasDataLayout`.

Reply:
hmm... maybe then just pull out `Scalar::Raw` arm into a private method?

Reply:
Tried that, at least for the part that can also be shared with `Debug::fmt`.